internal/faults/faults.go

```diff
@@ -0,0 +1,277 @@
+		Name: "faultbp_fault_requests_total",
+		Help: "Total count of requests seen by the fault injection middleware.",
+	}, []string{
+		clientNameLabel,
+		serviceLabel,
+		methodLabel,
+		protocolLabel,
+		successLabel,
+		delayInjectedLabel,
+		abortInjectedLabel,
+	})
+)
+
+// Headers is an interface to be implemented by the caller to allow
+// protocol-specific header lookup. Using an interface here rather than a
+// function type avoids any potential closure requirements of a function.
+type Headers interface {
+	// Lookup returns the value of a protocol-specific header with the
+	// given key.
+	Lookup(ctx context.Context, key string) (string, error)
+}
+
+// Resume is the function type to continue processing the protocol-specific
+// request without injecting a fault.
+type Resume[T any] func() (T, error)
+
+// Abort is the function type to inject a protocol-specific fault with the
+// given code and message.
+type Abort[T any] func(code int, message string) (T, error)
+
+// The canonical address for a cluster-local address is <service>.<namespace>,
+// without the local cluster suffix or port. The canonical address for a
+// non-cluster-local address is the full original address without the port.
+func getCanonicalAddress(serverAddress string) string {
+	// Cluster-local address.
+	if i := strings.Index(serverAddress, ".svc.cluster.local"); i != -1 {
+		return serverAddress[:i]
+	}
+	// External host:port address.
+	if i := strings.LastIndex(serverAddress, ":"); i != -1 {
+		port := serverAddress[i+1:]
+		// Verify this is actually a port number.
+		if port != "" && port[0] >= '0' && port[0] <= '9' {
+			return serverAddress[:i]
+		}
+	}
+	// Other address, i.e. unix domain socket.
+	return serverAddress
+}
+
+func parsePercentage(percentage string) (int, error) {
+	if percentage == "" {
+		return 100, nil
+	}
+	intPercentage, err := strconv.Atoi(percentage)
+	if err != nil {
+		return 0, fmt.Errorf("provided percentage %q is not a valid integer: %w", percentage, err)
+	}
+	if intPercentage < 0 || intPercentage > 100 {
+		return 0, fmt.Errorf("provided percentage \"%d\" is outside the valid range of [0-100]", intPercentage)
+	}
+	return intPercentage, nil
+}
+
+// Global variable to allow overriding in tests.
+var selected = func(percentage int) bool {
+	// Use a different random integer per feature as per
+	// https://github.com/grpc/proposal/blob/master/A33-Fault-Injection.md#evaluate-possibility-fraction.
+	return rand.IntN(100) < percentage
+}
+
+// Global variable to allow overriding in tests.
+var sleep = func(ctx context.Context, d time.Duration) error {
+	t := time.NewTimer(d)
+	select {
+	case <-t.C:
+	case <-ctx.Done():
+		t.Stop()
+		return ctx.Err()
+	}
+	return nil
+}
+
+// Injector contains the data common across all requests needed to inject
+// faults on outgoing requests.
+type Injector[T any] struct {
+	clientName, callerName     string
+	abortCodeMin, abortCodeMax int
```

Comment:
Formatting nit: each field on its own line unless I didn't see this in our style guide :eyes:

Reply:
oops! fixed